syntax = "proto3";
package Manager;

service Manager {
    rpc register(RegisterRequest) returns (ShardInfo) {}
    rpc guild_count(GuildCountRequest) returns (GuildInfo) {}
    rpc checkin(ShardID) returns (CheckInResponse) {}
    rpc publish_file(stream File) returns (Url) {}
    rpc all_guilds(AllGuildsRequest) returns (stream Guild) {}
    rpc guild_update(stream Guild) returns (UpdateResponse) {}
}

message RegisterRequest {
}

message GuildCountRequest {
}

message CheckInResponse {
}

message AllGuildsRequest {
}

message UpdateResponse {
}

message ShardInfo {
    uint32 shard_id = 1;
    uint32 shard_count = 2;
}

message GuildInfo {
    uint32 guild_count = 1;
    uint32 user_count = 2;
}

message ShardID {
    uint32 shard_id = 1;
}

message File {
    string location = 1;
    string name = 2;
    string filetype = 3;
    bytes file = 4;
}

message Url {
    string url = 1;
}

message Guild {
    uint32 shard_id = 1;
    fixed64 id = 2;
    string name = 3;
    string icon = 4;
    string splash = 5;
    fixed64 owner_id = 6;
    string region = 7;
    uint32 afk_timeout = 8;
    bool unavailable = 9;
    int32 max_members = 10;
    string banner = 11;
    string description = 12;
    int32 mfa_level = 13;
    uint32 premium_tier = 14;
    int32 premium_subscription_count = 15;
    string preferred_locale = 16;
    int32 member_count = 17;
    repeated fixed64 admin_ids = 18;
    repeated string features = 19;
}
